python/ray/rllib/agents/sac/sac_policy_graph.py

```diff
@@ -0,0 +1,233 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from numbers import Number
+
+import numpy as np
+import tensorflow as tf
+
+from gym.spaces import Box
+
+import ray
+from ray.rllib.utils.error import UnsupportedSpaceException
+from ray.rllib.evaluation.tf_policy_graph import TFPolicyGraph
+from ray.rllib.utils.annotations import override
+
+from .models import GaussianLatentSpacePolicy, feedforward_model
+
+
+class SACPolicyGraph(TFPolicyGraph):
+    def __init__(self, observation_space, action_space, config):
+        if not isinstance(action_space, Box):
+            # TODO(hartikainen): Should we support discrete action spaces?
+            # I've seen several people requesting support for it.
+            raise UnsupportedSpaceException(
+                "Action space {} is not supported for SAC.".format(
+                    action_space))
+
+        config = dict(ray.rllib.agents.sac.sac.DEFAULT_CONFIG, **config)
+
+        self.config = config
+
+        self.action_space = action_space
+        self.observation_space = observation_space
+
+        # create global step for counting the number of update operations
+        # self.session = self.sess = tf.keras.backend.get_session()
+        self.session = self.sess = tf.get_default_session()
+        self.session._graph = tf.get_default_graph()
+        tf.keras.backend.set_session(self.session)
```

Comment:
The keras and tensorflow sessions and graphs get messed up. This fixes the session problem (i.e. all the keras models use different session than the `tf.get_default_session()`) but the graph problems still persist. When I run the `global_variables_initializer()`, I get an error saying that none of the variables that are being initialized are not found in the graph. I'm not quite sure yet how to fix this.

Reply:
Weird, what's the error here? Is the graph definitely different from the default one (which is set in policy_evaluator.py:261:with tf.Graph().as_default()).

Reply:
Here's the error I currently get:
```python
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 300, in __init__                                                                                                                                                                                                            
    fetch, allow_tensor=True, allow_operation=True))                                                                                                                                                                                                                                                                                              
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3490, in as_graph_element                                                                                                                                                                                                    
    return self._as_graph_element_locked(obj, allow_tensor, allow_operation)                                                                                                                                                                                                                                                                      
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3569, in _as_graph_element_locked                                                                                                                                                                                            
    raise ValueError("Tensor %s is not an element of this graph." % obj)                                                                                                                                                                                                                                                                          
ValueError: Tensor Tensor("default/dense_3/kernel/Read/ReadVariableOp:0", shape=(3, 256), dtype=float32) is not an element of this graph.                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                                                  
During handling of the above exception, another exception occurred:                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                                                  
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                
  File "./test/test_sac.py", line 16, in testSAC                                                                                                                                                                                                                                                                                                  
    print(agent.train())                                                                                                                                                                                                                                                                                                                          
  File "/home/kristian/github/hartikainen/ray/python/ray/rllib/agents/agent.py", line 279, in train                                                                                                                                                                                                                                               
    result = Trainable.train(self)                                                                                                                                                                                                                                                                                                                
  File "/home/kristian/github/hartikainen/ray/python/ray/tune/trainable.py", line 147, in train                                                                                                                                                                                                                                                   
    result = self._train()                                                                                                                                                                                                                                                                                                                        
  File "/home/kristian/github/hartikainen/ray/python/ray/rllib/agents/sac/sac.py", line 175, in _train                                                                                                                                                                                                                                            
    self.update_Q_target_if_needed()                                                                                                                                                                                                                                                                                                              
  File "/home/kristian/github/hartikainen/ray/python/ray/rllib/agents/sac/sac.py", line 192, in update_Q_target_if_needed                                                                                                                                                                                                                         
    lambda p, _: p.update_target())                                                                                                                                                                                                                                                                                                               
  File "/home/kristian/github/hartikainen/ray/python/ray/rllib/evaluation/policy_evaluator.py", line 509, in foreach_trainable_policy                                                                                                                                                                                                             
    func(policy, pid) for pid, policy in self.policy_map.items()                                                                                                                                                                                                                                                                                  
  File "/home/kristian/github/hartikainen/ray/python/ray/rllib/evaluation/policy_evaluator.py", line 510, in <listcomp>                                                                                                                                                                                                                           
    if pid in self.policies_to_train                                                                                                                                                                                                                                                                                                              
  File "/home/kristian/github/hartikainen/ray/python/ray/rllib/agents/sac/sac.py", line 192, in <lambda>                                                                                                                                                                                                                                          
    lambda p, _: p.update_target())                                                                                                                                                                                                                                                                                                               
  File "/home/kristian/github/hartikainen/ray/python/ray/rllib/agents/sac/sac_policy_graph.py", line 228, in update_target                                                                                                                                                                                                                        
    source_params = self.Q.get_weights()                                                                                                                                                                                                                                                                                                          
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 483, in get_weights                                                                                                                                                                                                   
    return backend.batch_get_value(weights)                                                                                                                                                                                                                                                                                                       
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/keras/backend.py", line 2725, in batch_get_value                                                                                                                                                                                                     
    return get_session().run(tensors)                                                                                                                                                                                                                                                                                                             
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 929, in run                                                                                                                                                                                                                 
    run_metadata_ptr)                                                                                                                                                                                                                                                                                                                             
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1137, in _run                                                                                                                                                                                                               
    self._graph, fetches, feed_dict_tensor, feed_handles=feed_handles)                                                                                                                                                                                                                                                                            
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 471, in __init__                                                                                                                                                                                                            
    self._fetch_mapper = _FetchMapper.for_fetch(fetches)                                                                                                                                                                                                                                                                                          
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 261, in for_fetch                                                                                                                                                                                                           
    return _ListFetchMapper(fetch)                                                                                                                                                                                                                                                                                                                
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 370, in __init__                                                                                                                                                                                                            
    self._mappers = [_FetchMapper.for_fetch(fetch) for fetch in fetches]                                                                                                                                                                                                                                                                          
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 370, in <listcomp>                                                                                                                                                                                                          
    self._mappers = [_FetchMapper.for_fetch(fetch) for fetch in fetches]                                                                                                                                                                                                                                                                          
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 271, in for_fetch                                                                                                                                                                                                           
    return _ElementFetchMapper(fetches, contraction_fn)                                                                                                                                                                                                                                                                                           
  File "/home/kristian/conda/envs/softlearning/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 307, in __init__                                                                                                                                                                                                            
    'Tensor. (%s)' % (fetch, str(e)))                                                                                                                                                                                                                                                                                                             
ValueError: Fetch argument <tf.Variable 'default/dense_3/kernel:0' shape=(3, 256) dtype=float32> cannot be interpreted as a Tensor. (Tensor Tensor("default/dense_3/kernel/Read/ReadVariableOp:0", shape=(3, 256), dtype=float32) is not an element of this graph.)
```

Depending on how I change the sessions here, I might get some other similar errors. I'll see what the correct way of handling sessions/graphs with keras models is.